neqo-http3/src/recv_message.rs

```diff
@@ -377,10 +377,10 @@ impl RecvMessage {
         }
     }
 
-    fn headers_valid(&self, headers: &[Header]) -> Res<bool> {
+    fn sanitize_headers(&self, headers: &mut Vec<Header>) -> Res<()> {
```

Comment:
I would create a new function for the stuff you extracted out.  Then this function doesn't need to take a mutable value.

You could even move the validation to `Header::is_allowed` and just call: `headers.retain(Header::is_allowed)` in `add_headers` after validating.